src/librustc/mir/tcx.rs

```diff
@@ -151,12 +153,124 @@ impl<'tcx> Place<'tcx> {
                     }
                 },
                 _ => None,
-            }
+            }
             _ => None,
         }
     }
 }
 
+impl<'tcx> PlaceBase<'tcx> {
+    pub fn ty(&self, local_decls: &impl HasLocalDecls<'tcx>) -> Ty<'tcx> {
+        match self {
+            PlaceBase::Local(index) => local_decls.local_decls()[*index].ty,
+            PlaceBase::Promoted(data) => data.1,
+            PlaceBase::Static(data) => data.ty,
+        }
+    }
+}
+
+impl<'tcx> NeoPlace<'tcx> {
+    pub fn ty<'a, 'gcx>(
+        &self,
+        local_decls: &impl HasLocalDecls<'tcx>,
+        tcx: TyCtxt<'a, 'gcx, 'tcx>,
+    ) -> PlaceTy<'tcx> {
+        // the PlaceTy is the *final* type with all projection applied
+        // if there is no projection, that just refers to `base`:
+        //
+        // Place: base.[a, b, c]
+        //                    ^-- projection
+        //                    ^-- PlaceTy
+        //
+        // Place: base.[]
+        //             ^^-- no projection
+        //        ^^^^-- PlaceTy
+
+        let mut place_ty = PlaceTy::from(self.base.ty(local_decls));
+
+        // apply .projection_ty() to all elems but only returns the final one.
+        for elem in self.elems.iter() {
+            place_ty = place_ty.projection_ty(tcx, elem);
+        }
+
+        place_ty
+    }
+
+    /// If this is a field projection, and the field is being projected from a closure type,
+    /// then returns the index of the field being projected. Note that this closure will always
+    /// be `self` in the current MIR, because that is the only time we directly access the fields
+    /// of a closure type.
+    pub fn is_upvar_field_projection<'cx, 'gcx>(
+        &self,
+        mir: &'cx Mir<'tcx>,
+        tcx: &TyCtxt<'cx, 'gcx, 'tcx>,
+    ) -> Option<Field> {
+        let (place, by_ref) = if let Some(ProjectionElem::Deref) = self.elems.last() {
+            (self.projection_base(*tcx), true)
```

Comment:
You can rewrite this as going *backwards* in the projection list.

If you want, you can use `let mut elems = self.elems.iter().rev();` and then call `next()` on that once, check whether it's `Deref`, and if it is, call `next()` again.

You can also use indices.

Reply:
It feels like this is a case where treating the projection as a tree (sort of what this code is doing) is a bit cleaner. In particular, we want to look for something like `P.f` or `*P.f` and then test the type of `P`. We can certainly "walk backwards" over the elements, but once we've peeled off the `*` and the `.f` we sitll kind of want to extract the `P` as a unit, right?